lib/ansible/modules/cloud/azure/azure_rm_networkinterface.py

```diff
@@ -227,53 +247,78 @@
+
+
 def nic_to_dict(nic):
-    result = dict(
+    ip_configurations = [
+        dict(
+            name=config.name,
+            private_ip_address=config.private_ip_address,
+            private_ip_allocation_method=config.private_ip_allocation_method,
+            subnet=subnet_to_dict(config.subnet),
+            primary=config.primary,
+            public_ip_address=dict(
+                id=config.public_ip_address.id,
+                name=azure_id_to_dict(config.public_ip_address.id).get('publicIPAddresses'),
+                public_ip_allocation_method=config.public_ip_address.public_ip_allocation_method
+            ) if config.public_ip_address else None
+        ) for config in nic.ip_configurations
+    ]
+    return dict(
         id=nic.id,
         name=nic.name,
         type=nic.type,
         location=nic.location,
         tags=nic.tags,
-        network_security_group=dict(),
-        ip_configuration=dict(
-            name=nic.ip_configurations[0].name,
-            private_ip_address=nic.ip_configurations[0].private_ip_address,
-            private_ip_allocation_method=nic.ip_configurations[0].private_ip_allocation_method,
-            subnet=dict(),
-            public_ip_address=dict(),
-        ),
+        network_security_group=dict(
+            id=nic.network_security_group.id,
+            name=azure_id_to_dict(nic.network_security_group.id).get('networkSecurityGroups')
+        ) if nic.network_security_group else None,
         dns_settings=dict(
             dns_servers=nic.dns_settings.dns_servers,
             applied_dns_servers=nic.dns_settings.applied_dns_servers,
             internal_dns_name_label=nic.dns_settings.internal_dns_name_label,
             internal_fqdn=nic.dns_settings.internal_fqdn
         ),
+        ip_configurations=ip_configurations,
+        ip_configuration=ip_configurations[0] if len(ip_configurations) == 1 else None,  # for compatiable issue, keep this field
         mac_address=nic.mac_address,
-        primary=nic.primary,
         enable_ip_forwarding=nic.enable_ip_forwarding,
         provisioning_state=nic.provisioning_state,
         etag=nic.etag,
     )
 
-    if nic.network_security_group:
-        result['network_security_group']['id'] = nic.network_security_group.id
-        id_keys = azure_id_to_dict(nic.network_security_group.id)
-        result['network_security_group']['name'] = id_keys['networkSecurityGroups']
 
-    if nic.ip_configurations[0].subnet:
-        result['ip_configuration']['subnet']['id'] = \
-            nic.ip_configurations[0].subnet.id
-        id_keys = azure_id_to_dict(nic.ip_configurations[0].subnet.id)
-        result['ip_configuration']['subnet']['virtual_network_name'] = id_keys['virtualNetworks']
-        result['ip_configuration']['subnet']['name'] = id_keys['subnets']
+def construct_ip_configuration_set(raw):
+    configurations = [str(dict(
+        private_ip_allocation_method=item.get('private_ip_allocation_method').encode('ascii'),
```

Comment:
`_text.py` has a `to_bytes`, `to_text` and `to_native` methods which should be used instead of `encode` and `decode`. https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/_text.py

`to_bytes` will encode the string as a bytes string
`to_text` will encode the string as a unicode string
`to_native` will encode the string as a text string, this calls to_bytes for Python 2 and to_text for Python 3